### PR TITLE
Drop Shadow Improvements

### PR DESCRIPTION
**Pull request by AidanCaruso**
## Summary

This PR is aimed at improving the current drop shadow implementation and integrates it properly into the rendering refactor.

## References

closes #_insert number here_

## Checklist
- [ ] If this PR is still a WIP, convert to a draft
- [ ] When this PR is ready, mark it as "Ready for review"
- [ ] [ensure all checks pass](https://github.com/XRFoundation/XREngine/wiki/Testing-&-Contributing)
- [ ] Changes have been manually QA'd
- [ ] Changes reviewed by at least 2 approved reviewer

## QA Steps

_List any additional steps required to QA the changes of this PR, as well as any supplemental images or videos._

### CODE REVIEW

File: packages/engine/src/scene/systems/ShadowSystem.ts [hunk 1]
```diff
@@ -163,43 +171,36 @@ export default async function ShadowSystem(world: World) {
...
     dropShadows.instanceMatrix.needsUpdate = true
   }
 
   const execute = () => {
-    const useShadows = getShadowsEnabled()
-    if (!useShadows) {
-      CreateDropShadows()
-      return
-    }
+    UpdateDropShadows()
```
Comment (HexaField) [inline packages/engine/src/scene/systems/ShadowSystem.ts:203]: should start lower case

File: packages/engine/src/scene/systems/ShadowSystem.ts [hunk 2]
```diff
@@ -142,18 +143,25 @@ export default async function ShadowSystem(world: World) {
   let dropShadows = new InstancedMesh(shadowGeometry, shadowMaterial, 0)
   dropShadows.matrixAutoUpdate = false
 
-  const dropShadowReactor = startQueryReactor([DropShadowComponent, GroupComponent], function modifyShadowCount() {
+  const dropShadowReactor = startQueryReactor([DropShadowComponent, GroupComponent], function (props) {
+    if (getShadowsEnabled()) return null
+
     world.scene.remove(dropShadows)
```
Comment (HexaField) [inline packages/engine/src/scene/systems/ShadowSystem.ts:149]: logic here should be in a `useEffect(() => ` with the dependencies being whatever this relies on (drop shadow radius, group component length etc)

**Review by HexaField — CHANGES_REQUESTED** [review]

File: packages/engine/src/scene/systems/ShadowSystem.ts [hunk 3]
```diff
@@ -142,18 +143,25 @@ export default async function ShadowSystem(world: World) {
...
+    const minRadius = 0.15
+    new Box3().setFromObject(getComponent(props.root.entity, GroupComponent)[0]).getBoundingSphere(sphere)
+    getComponent(props.root.entity, DropShadowComponent).radius = Math.max(sphere.radius * 2, minRadius)
     return null
   })
 
   let sceneObjects = Array.from(Engine.instance.currentWorld.objectLayerList[ObjectLayers.Camera] || [])
 
-  const CreateDropShadows = () => {
+  const UpdateDropShadows = () => {
```
Comment (HexaField) [inline packages/engine/src/scene/systems/ShadowSystem.ts:164]: let's rename this to `updateDropShadows` to keep variable names consistently camelCase across the codebase

File: packages/engine/src/scene/systems/ShadowSystem.ts [hunk 4]
```diff
@@ -163,43 +171,36 @@ export default async function ShadowSystem(world: World) {
...
-      new Box3().setFromObject(group[0], false).getBoundingSphere(sphere)
       const distanceShrinkBias = 3
-      const sizeBias = 1.5
-      const finalSize = sphere.radius * Math.min(distanceShrinkBias / intersects[0].distance, 1) * sizeBias
+      const sizeBias = 1
+      const finalSize = dropShadowComponent.radius * Math.min(distanceShrinkBias / intersected.distance, 1) * sizeBias
 
-      let shadowMatrix = new Matrix4()
-      const shadowRotation = new Quaternion().setFromUnitVectors(intersects[0].face.normal, V_001)
+      const shadowRotation = new Quaternion().setFromUnitVectors(intersected.face.normal, V_001)
```
Comment (HexaField) [inline packages/engine/src/scene/systems/ShadowSystem.ts:192]: any `new` math objects being created often should instead be created once in module scope and reused for all repetitions

**Review by HexaField — CHANGES_REQUESTED** [review]

File: packages/engine/src/scene/systems/ShadowSystem.ts (hunk 3 above)
Comment (speigg) [reply packages/engine/src/scene/systems/ShadowSystem.ts:164]: Is this the entire system logic? Let's just move it directly into the system execute function

**Review by speigg — CHANGES_REQUESTED** [review]

File: packages/engine/src/scene/systems/ShadowSystem.ts [hunk 5]
```diff
@@ -50,6 +50,29 @@ import { ObjectLayers } from '../constants/ObjectLayers'
...
+
+const shadowGeometry = new PlaneGeometry(1, 1, 1, 1)
+const shadowMaterial = new MeshBasicMaterial({
+  side: DoubleSide,
+  transparent: true,
+  depthTest: true,
+  depthWrite: false
+})
+
+AssetLoader.loadAsync(`${config.client.fileServer}/projects/default-project/public/drop-shadow.png`).then(
```
Comment (HexaField) [inline packages/engine/src/scene/systems/ShadowSystem.ts:65]: asset loading should not be done in module scope

File: packages/engine/src/scene/systems/ShadowSystem.ts [hunk 6]
```diff
@@ -125,43 +148,30 @@ export default async function ShadowSystem(world: World) {
...
+      dropShadows = new InstancedMesh(shadowGeometry, shadowMaterial, shadowComponentQuery().length)
+      dropShadows.matrixAutoUpdate = false
+      dropShadows.layers.disable(ObjectLayers.Camera)
+      world.scene.add(dropShadows)
+
+      const sphere = new Sphere()
+      const minRadius = 0.15
+      new Box3().setFromObject(groupComponent[0]).getBoundingSphere(sphere)
+      dropShadowComponent.radius = Math.max(sphere.radius * 2, minRadius)
+    }, [dropShadowComponent.radius, groupComponent.length, dropShadows])
```
Comment (HexaField) [inline packages/engine/src/scene/systems/ShadowSystem.ts:169]: dropShadowComponent.radius should not be a dependency of an effect that mutates dropShadowComponent.radius

**Review by HexaField — CHANGES_REQUESTED** [review]

File: packages/engine/src/scene/systems/ShadowSystem.ts [hunk 7]
```diff
@@ -124,36 +146,32 @@ export default async function ShadowSystem(world: World) {
...
-    dropShadows.matrixAutoUpdate = false
-    dropShadows.layers.disable(ObjectLayers.Camera)
-    world.scene.add(dropShadows)
     return null
   })
 
-  let sceneObjects = Array.from(Engine.instance.currentWorld.objectLayerList[ObjectLayers.Camera] || [])
-
-  const CreateDropShadows = () => {
+  const execute = () => {
```
Comment (HexaField) [inline packages/engine/src/scene/systems/ShadowSystem.ts:174]:
```ts
const useShadows = getShadowsEnabled()
if (!useShadows) {
```

This still needs to be the condition for the drop shadow functionality

**Review by HexaField — CHANGES_REQUESTED** [review]

File: packages/engine/src/scene/systems/ShadowSystem.ts (hunk 7 above)
Comment (AidanCaruso) [reply packages/engine/src/scene/systems/ShadowSystem.ts:174]: It is in the case of the reactor, I just forgot to reinclude it in execute.

File: packages/engine/src/scene/systems/ShadowSystem.ts [hunk 8]
```diff
@@ -142,62 +144,74 @@ export default async function ShadowSystem(world: World) {
...
-  })
+  const shadowComponentQuery = defineQuery([DropShadowComponent, GroupComponent])
+
+  const shadowOffset = new Vector3(0, 0.01, 0)
 
   let sceneObjects = Array.from(Engine.instance.currentWorld.objectLayerList[ObjectLayers.Camera] || [])
 
-  const CreateDropShadows = () => {
+  const dropShadowReactor = startQueryReactor([DropShadowComponent, GroupComponent], function (props) {
+    if (getShadowsEnabled()) return null
```
Comment (HexaField) [inline packages/engine/src/scene/systems/ShadowSystem.ts:154]: Any if statements must be in the effect. Having a differing number of effects in a reactor will cause bad errors in react.

**Review by HexaField — CHANGES_REQUESTED** [review]

File: packages/engine/src/scene/systems/ShadowSystem.ts (hunk 8 above)
Comment (AidanCaruso) [reply packages/engine/src/scene/systems/ShadowSystem.ts:154]: Got it, going to read up on this.

**Review by HexaField — CHANGES_REQUESTED** [review]
inside the main body of the reactor (not the effect) should be useComponent instead of getComponent, such that we can reactively respond to component changes. it would be wise to also add dependencies of getShadowsEnabled or implement a separate useShadowsEnabled to reactively respond to those state changes

File: packages/engine/src/scene/systems/ShadowSystem.ts [hunk 9]
```diff
@@ -163,15 +164,15 @@ export default async function ShadowSystem(world: World) {
...
         const sphere = new Sphere()
-        new Box3().setFromObject(groupComponent[0]).getBoundingSphere(sphere)
-        dropShadowComponent.radius = Math.max(sphere.radius * 2, minRadius)
-        dropShadowComponent.center = groupComponent[0].worldToLocal(sphere.center)
+        new Box3().setFromObject(groupComponent.value[0]).getBoundingSphere(sphere)
+        dropShadowComponent.radius.set(Math.max(sphere.radius * 2, minRadius))
+        dropShadowComponent.center.set(groupComponent.value[0].worldToLocal(sphere.center))
       }
-    }, [groupComponent.length, dropShadows])
+    }, [groupComponent.length, getShadowsEnabled()])
```
Comment (HexaField) [inline packages/engine/src/scene/systems/ShadowSystem.ts:175]: this won't work, let's just add alongside the getShadowsEnabled defintion

```ts

export const useShadowsEnabled = () => {
  const engineRendererState = useHookstate(getState(EngineRendererState))
  return (
    !isHeadset() && engineRendererState.useShadows.value && engineRendererState.renderMode.value === RenderModes.SHADOW
  )
}
```

**Review by HexaField — CHANGES_REQUESTED** [review]

File: packages/engine/src/scene/systems/ShadowSystem.ts [hunk 10]
```diff
@@ -172,7 +172,7 @@ export default async function ShadowSystem(world: World) {
         dropShadowComponent.radius.set(Math.max(sphere.radius * 2, minRadius))
         dropShadowComponent.center.set(groupComponent.value[0].worldToLocal(sphere.center))
       }
-    }, [groupComponent.length, getShadowsEnabled()])
+    }, [groupComponent.length, useShadowsEnabled()])
```
Comment (HexaField) [inline packages/engine/src/scene/systems/ShadowSystem.ts:175]: let's pull this call out of the dep and have it with the other ones above the effect

**Review by HexaField — CHANGES_REQUESTED** [review]

File: packages/engine/src/scene/systems/ShadowSystem.ts (hunk 10 above)
Comment (AidanCaruso) [reply packages/engine/src/scene/systems/ShadowSystem.ts:175]: Ah yes, sorry my mistake. I was rushing so I didn't give enough thought to the last commits.

File: packages/engine/src/scene/systems/ShadowSystem.ts [hunk 11]
```diff
@@ -142,62 +144,85 @@ export default async function ShadowSystem(world: World) {
...
+
+  let sceneObjects = Array.from(Engine.instance.currentWorld.objectLayerList[ObjectLayers.Camera] || [])
+
+  const dropShadowReactor = startQueryReactor([DropShadowComponent, GroupComponent], function (props) {
+    const dropShadowComponent = useComponent(props.root.entity, DropShadowComponent)
+    const groupComponent = useComponent(props.root.entity, GroupComponent)
+    const useShadows = useShadowsEnabled()
+
+    useEffect(() => {
+      const getShadows = getShadowsEnabled()
```
Comment (HexaField) [inline packages/engine/src/scene/systems/ShadowSystem.ts:157]: No need to have an extra call here, we can use our previously define useShadows instead

**Review by HexaField — CHANGES_REQUESTED** [review]

File: packages/engine/src/scene/components/ShadowComponent.ts [hunk 12]
```diff
@@ -50,10 +52,9 @@ export const ShadowComponent = defineComponent({
 })
 
 function ShadowReactor({ root }: EntityReactorProps) {
+  const shadowComponent = useOptionalComponent(root.entity, ShadowComponent)
```
Comment (AidanCaruso) [inline packages/engine/src/scene/components/ShadowComponent.ts:55]: I switched to useOptionalComponent before the effect, is this best practice?

Comment (HexaField) [reply packages/engine/src/scene/components/ShadowComponent.ts:55]: perfect!

**Review by speigg — APPROVED** [review]

**Review by HexaField — APPROVED** [review]

**Pull request merged**